src/lib/branchpredictors/TagePredictor.cc

```diff
@@ -0,0 +1,279 @@
+#include "simeng/branchpredictors/TagePredictor.hh"
+
+namespace simeng {
+
+TagePredictor::TagePredictor(ryml::ConstNodeRef config)
+    : btbBits_(config["Branch-Predictor"]["BTB-Tag-Bits"].as<uint8_t>()),
+      tageTableBits_(
+          config["Branch-Predictor"]["Tage-Table-Bits"].as<uint8_t>()),
+      numTageTables_(
+          config["Branch-Predictor"]["Num-Tage-Tables"].as<uint8_t>()),
+      satCntBits_(
+          config["Branch-Predictor"]["Saturating-Count-Bits"].as<uint8_t>()),
+      globalHistoryLength_(
+          config["Branch-Predictor"]["Global-History-Length"].as<uint16_t>()),
+      rasSize_(config["Branch-Predictor"]["RAS-entries"].as<uint16_t>()),
+      globalHistory_(1 << (numTageTables_ + 1)),
+      tagLength_(config["Branch-Predictor"]["Tag-Length"].as<uint8_t>()) {
+  // Calculate the saturation counter boundary between weakly taken and
+  // not-taken. `(2 ^ num_sat_cnt_bits) / 2` gives the weakly taken state
+  // value
+  uint8_t weaklyTaken = (uint8_t)1 << (satCntBits_ - 1);
+  uint8_t satCntVal = (config["Branch-Predictor"]["Fallback-Static-Predictor"]
+                           .as<std::string>() == "Always-Taken")
+                          ? weaklyTaken
+                          : (weaklyTaken - 1);
+
+  // Set up non-tagged default prediction table
+  btb_ = std::vector<std::pair<uint8_t, uint64_t>>(1ul << btbBits_,
+                                                   {satCntVal, 0});
+
+  // Set up tagged prediction tables
+  for (uint32_t i = 0; i < numTageTables_; i++) {
+    std::vector<TageEntry> newTable;
+    for (uint32_t j = 0; j < (1ul << tageTableBits_); j++) {
+      TageEntry newEntry = {2, 0, 1, 0};
```

Comment:
Would we not want to initialise the `TageEntry` with a SatCnt equal to the once used in the `btb_`?

Reply:
Yep, good catch